src/libraries/Common/tests/System/ModifiedTypeTests.cs

```diff
@@ -0,0 +1,422 @@
+
+using System.Reflection;
+using System.Reflection.Tests;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using Xunit;
+
+namespace System.Tests.Types
+{
+    public partial class ModifiedTypeTests
+    {
+        private const BindingFlags Bindings = BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.Static | BindingFlags.DeclaredOnly;
+
+        [Fact]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/71095", TestRuntimes.Mono)]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/71883", typeof(PlatformDetection), nameof(PlatformDetection.IsNativeAot))]
+        public static unsafe void TestFields_Modified()
+        {
+            Type volatileInt = typeof(ModifiedTypeHolder).Project().GetField(nameof(ModifiedTypeHolder._volatileInt), Bindings).GetModifiedFieldType();
+            Verify(volatileInt);
+
+            Type volatileIntArray = typeof(ModifiedTypeHolder).Project().GetField(nameof(ModifiedTypeHolder._volatileIntArray), Bindings).GetModifiedFieldType();
+            Verify(volatileIntArray);
+
+            Type volatileIntPointer = typeof(ModifiedTypeHolder).Project().GetField(nameof(ModifiedTypeHolder._volatileIntPointer), Bindings).GetModifiedFieldType();
+            Verify(volatileIntPointer);
+            Type volatileIntPointerElementType = volatileIntPointer.GetElementType();
+            Assert.True(IsModifiedType(volatileIntPointerElementType));
+            Assert.True(ReferenceEquals(volatileIntPointerElementType.UnderlyingSystemType, typeof(int).Project()));
+            Assert.Equal(0, volatileIntPointerElementType.UnderlyingSystemType.GetRequiredCustomModifiers().Length);
+            Assert.Equal(0, volatileIntPointerElementType.UnderlyingSystemType.GetOptionalCustomModifiers().Length);
+
+            Type volatileFcnPtr = typeof(ModifiedTypeHolder).Project().GetField(nameof(ModifiedTypeHolder._volatileFcnPtr), Bindings).GetModifiedFieldType();
+            Verify(volatileFcnPtr);
+            Assert.True(IsModifiedType(volatileFcnPtr.GetFunctionPointerReturnType()));
+            Assert.Equal(1, volatileFcnPtr.GetFunctionPointerParameterTypes().Length);
+            Assert.True(IsModifiedType(volatileFcnPtr.GetFunctionPointerParameterTypes()[0]));
+
+            void Verify(Type type)
+            {
+                Assert.True(IsModifiedType(type));
+
+                Assert.Equal(1, type.GetRequiredCustomModifiers().Length);
+                Assert.Equal(typeof(IsVolatile).Project(), type.GetRequiredCustomModifiers()[0]);
+
+                Assert.Equal(0, type.GetOptionalCustomModifiers().Length);
+
+                Assert.Equal(0, type.UnderlyingSystemType.GetRequiredCustomModifiers().Length);
+                Assert.Equal(0, type.UnderlyingSystemType.GetOptionalCustomModifiers().Length);
+            }
+        }
+
+        // NOTE: commented out due to compiler issue on NativeAOT:
```

Comment:
Once CI is green, I'll move these to a non-shared file and log an issue for NativeAot. Basically it doesn't like `ref bool` for a function pointer paramter.

Reply:
> I'll move these to a non-shared file and log an issue for NativeAot. 

I do not think that it is going to help. We use one test binaries across all runtimes. The build system does not have a provision to use runtime-specific conditions in the test compilation.

> it doesn't like ref bool for a function pointer parameter.

That should not be hard to fix.

Reply:
I'll log a bug.

Reply:
See https://github.com/dotnet/runtime/issues/81117